usb: warn about too small REQUEST_BUFFER

jfischer-no opened the pull request:
Warn about too small REQUEST_BUFFER.
Clarify usb_handle_request function parameter.

jfischer-no committed:
usb: terminate if the REQUEST_BUFFER is too small

If the CONFIG_USB_REQUEST_BUFFER_SIZE is too small,
the data will be truncated and the device will not be
fully recognized by the host.
Terminate if the CONFIG_USB_REQUEST_BUFFER_SIZE is too small.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

finikorg commented on subsys/usb/usb_device.c:
```diff
@@ -232,9 +232,11 @@ static bool usb_handle_request(struct usb_setup_packet *setup,
  * @return N/A
  */
-static void usb_data_to_host(u16_t len)
+static void usb_data_to_host(u16_t rlen)
```
This does not belong to this patch and not needed IMO

jfischer-phytec-iot replied:
I will move it to another PR. But it is necessary. There are many `len`, can you say without much thought what is meant by `len` in the context?

finikorg replied:
At least I know that it is len, rlen is not better here.

carlescufi commented on subsys/usb/usb_device.c:
```diff
@@ -332,6 +332,10 @@ static void usb_handle_control_transfer(u8_t ep,
 		usb_dev.data_buf_residue = MIN(usb_dev.data_buf_len, length);
+		if (usb_dev.data_buf_len > CONFIG_USB_REQUEST_BUFFER_SIZE) {
+			LOG_ERR("REQUEST_BUFFER too small");
+		}
```
Can we add a call to `k_oops()` here so that the error is unconditionally detected

jfischer-phytec-iot replied:
done

jfischer-no committed:
samples: usb: increase REQUEST_BUFFER for samples with larger descriptor

Increase REQUEST_BUFFER for samples with larger descriptor.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
samples: hid-cdc: relax "Wait for DTR" while loop

Relax "Wait for DTR" while loop and give CPU resources
to low priority threads like logging.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

emob-nordic commented on subsys/usb/usb_device.c:
```diff
@@ -332,6 +332,11 @@ static void usb_handle_control_transfer(u8_t ep,
 		/* Send smallest of requested and offered length */
 		usb_dev.data_buf_residue = MIN(usb_dev.data_buf_len, length);
+		if (usb_dev.data_buf_len > CONFIG_USB_REQUEST_BUFFER_SIZE) {
+			LOG_ERR("REQUEST_BUFFER too small");
```
I the same function above
https://github.com/zephyrproject-rtos/zephyr/blob/e6dc03550cd934e18a6748f437483be82f14c4b9/subsys/usb/usb_device.c#L306
there is similar logging based on wrong buffer length. It's done based on 'Data transfer direction' but I am wondering if its possible to concatenate this into one LOG_ERR to make it more clear? 

What is more. Is it really an error? Refering to 9.3.5 of usb 2.0 spec:

> On an input request, a device must never return more data than is indicated by the wLength value; it may
return less. On an output request, wLength will always indicate the exact amount of data to be sent by the
host. Device behavior is undefined if the host should send more data than is specified in wLength.

emob-nordic commented on samples/subsys/usb/hid-cdc/prj.conf:
```diff
@@ -14,6 +14,7 @@ CONFIG_USB_DEVICE_LOG_LEVEL_INF=y
 
 CONFIG_USB_CDC_ACM=y
 CONFIG_USB_CDC_ACM_DEVICE_COUNT=2
+CONFIG_USB_REQUEST_BUFFER_SIZE=192
```
Why is size of 192 needed?

emob-nordic commented on samples/subsys/usb/cdc_acm_composite/prj.conf:
```diff
@@ -12,6 +12,7 @@ CONFIG_USB_COMPOSITE_DEVICE=y
 CONFIG_USB_CDC_ACM=y
 CONFIG_USB_CDC_ACM_DEVICE_COUNT=2
 CONFIG_USB_CDC_ACM_RINGBUF_SIZE=512
+CONFIG_USB_REQUEST_BUFFER_SIZE=192
```
Why is size of 192 needed?

jfischer-phytec-iot commented:
It is not quite right, I have overlooked some passages. In general it is difficult to control what happens with `usb_dev.data_buf `and `usb_dev.req_data`.
